### ISSUE TITLE
Forward-slash and space munging on Windows

### ISSUE THREAD

**Issue by mqudsi** [mesonbuild/meson#3565]
Running meson 0.46.0 with Python 3.7 under Windows 10 in either a VS developer command prompt or a standard command prompt with a basic project file (single-cpp `executable` and nothing more) returns the following:

```cmd
D:\GIT\RegistryRocket>meson build
The Meson build system
Version: 0.46.0
Source dir: D:\GIT\RegistryRocket
Build dir: D:\GIT\RegistryRocket\build
Build type: native build
Project name: RegistryRocket

meson.build:1:0: ERROR:  Unknown compiler(s): [['c:Program', 'Files', '(x86)Microsoft', 'Visual', 'Studio', '15.0VCbincl.exe']]
The follow exceptions were encountered:
Running "c:Program Files (x86)Microsoft Visual Studio 15.0VCbincl.exe --version" gave "[WinError 2] The system cannot find the file specified"

A full log can be found at D:\GIT\RegistryRocket\build\meson-logs\meson-log.txt
```

Forward slashes as path separators have not been processed literally, and spaces in the compiler path have been taken as token separators.

**Comment by nirbheek** [comment]
This setup is known to work well, so I am confused as to how you are running into this issue.

How did you install meson? Are you using the MSI installer, pip, or something else?

**Comment by mqudsi** [comment]
With Python 3.7b:

```
python -m pip install meson
```

And I created a file `meson.cmd` in my `%PATH%` with the following:

```batch
@python C:\Users\Mahmoud\AppData\Local\Programs\Python\Python37\Scripts\meson.py %*
```

thereafter executed

```batch
meson build
```

**Comment by mqudsi** [comment]
Do you have any thoughts? Is there a workaround?

**Comment by mqudsi** [comment]
Hey @jibsen thanks for the comment.

Unfortunately that didn't work.

```batch
D:\GIT\RegistryRocket>py %localappdata%\Programs\Python\Python37\Scripts\meson.py build
The Meson build system
Version: 0.46.0
Source dir: D:\GIT\RegistryRocket
Build dir: D:\GIT\RegistryRocket\build
Build type: native build
Project name: RegistryRocket

meson.build:1:0: ERROR:  Unknown compiler(s): [['C:Program', 'Files', '(x86)MicrosoftVisual', 'Studio', '15.0VCbincl.exe']]
The follow exceptions were encountered:
Running "C:Program Files (x86)MicrosoftVisual Studio 15.0VCbincl.exe --version" gave "[WinError 2] The system cannot find the file specified"

A full log can be found at D:\GIT\RegistryRocket\build\meson-logs\meson-log.txt

D:\GIT\RegistryRocket>
```

I get the same behavior in a vcvarsall shell and in a regular cmd prompt, so it seems that meson _is_ correctly reading `%CC%`, just parsing it wrong.

**Comment by mqudsi** [comment]
The problem is that in `environment.py`, the check `elif evar in os.environ` comes before the default handler. If `CC` or `CXX` is exported, the value of that variable is split on whitespace and each element is presumed to be a compiler possibility.

However, as is the default on Windows, if CC/CXX are not defined, that code is bypassed and the default compiler logic is reached instead.

Basically, `shlex.split` needs to handle `os='nt'`

**Comment by NickeZ** [comment]
I made an attempt at fixing the problem. Works for me on windows.

**Comment by mqudsi** [comment]
I imagine that won't work on paths that contain multiple consecutive spaces? Splitting should just be skipped entirely rather than attempting to recompose the parts, imho.

**Comment by NickeZ** [comment]
Good point, we stlil need to detect 'ccache gcc' but maybe it is easier to simply do `str.find('ccache')`...

**Comment by nirbheek** [comment]
> Windows can't execute .py files directly in the terminal

If you install Python with the right options enabled, it does allow you to execute .py files directly in the terminal. You have to enable the options that say "install for all users" and "add it to PATH", then install meson with pip and then typing `meson` in cmd.exe will execute meson.

**Comment by nirbheek** [comment]
> I don't think meson will pick up MSVC correctly from the CC environment variable, but I could be wrong.

You can't set `CC` and expect MSVC to be picked up properly on Windows. MSVC requires a bunch of other variables set to find the Windows SDK. The compiler might work, but you won't be able to compile any non-trivial programs.

What you want is to either use the visual studio command line prompt or [fix this issue](https://github.com/mesonbuild/meson/issues/2622).

**Comment by nirbheek** [comment]
> I get the same behavior in a vcvarsall shell and in a regular cmd prompt, so it seems that meson is correctly reading %CC%, just parsing it wrong.

When you're inside a vcvarsall shell, you don't need to set `CC` since if `cl.exe` is in `%PATH%` it should be picked up automatically. If it isn't, that's a bug and we should fix that.
